File: Client/file-exploder/Models/RemoteFile.swift
import Foundation

struct RemoteFile: Identifiable, Hashable {
    /// A byte-exact stand-in for `path`.
    ///
    /// Swift compares Strings by Unicode canonical equivalence, so an NFC and
    /// an NFD spelling of the same name are `==` and hash alike. A Linux server
    /// keeps them as two genuinely different files — a directory shared with a
    /// Mac routinely ends up holding both — and every identity check in the file
    /// list runs through this id: Table's row identity, the selection set, and
    /// `first(where: { $0.id == fileId })` behind the context menu. Sharing an
    /// id there means an action can land on the wrong file. Encoding the path's
    /// bytes keeps the id ASCII, so String equality is byte equality again.
    let id: String
    let name: String
    let path: String
    let size: Int64
    let modificationDate: Date
    let isDirectory: Bool
    let isSymlink: Bool
    let permissions: FilePermissions

    init(
        name: String,
        path: String,
        size: Int64,
        modificationDate: Date,
        isDirectory: Bool,
        isSymlink: Bool,
        permissions: FilePermissions
    ) {
        self.id = RemoteFile.identity(for: path)
        self.name = name
        self.path = path
        self.size = size
        self.modificationDate = modificationDate
        self.isDirectory = isDirectory
        self.isSymlink = isSymlink
        self.permissions = permissions
    }

    private static let hexDigits = Array("0123456789abcdef")

    private static func identity(for path: String) -> String {
        var characters: [Character] = []
        characters.reserveCapacity(path.utf8.count * 2)
        for byte in path.utf8 {
            characters.append(hexDigits[Int(byte >> 4)])
            characters.append(hexDigits[Int(byte & 0x0f)])
        }
        return String(characters)
    }

    var displayName: String {
        name
    }
    
    var formattedSize: String {
        if isDirectory {
            return "--"
        }
        return FormatUtils.formattedSize(size)
    }
    
    var formattedDate: String {
        FormatUtils.formattedDate(modificationDate)
    }

    var symbolicPermissions: String {
        permissions.symbolicString(isDirectory: isDirectory, isSymlink: isSymlink)
    }
    
    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }
    
    static func == (lhs: RemoteFile, rhs: RemoteFile) -> Bool {
        lhs.id == rhs.id
    }
}

struct FilePermissions: Hashable {
    let ownerRead: Bool
    let ownerWrite: Bool
    let ownerExecute: Bool
    let groupRead: Bool
    let groupWrite: Bool
    let groupExecute: Bool
    let otherRead: Bool
    let otherWrite: Bool
    let otherExecute: Bool
    
    func symbolicString(isDirectory: Bool, isSymlink: Bool) -> String {
        var str: String
        if isSymlink {
            str = "l"
        } else if isDirectory {
            str = "d"
        } else {
            str = "-"
        }
        
        str += ownerRead ? "r" : "-"
        str += ownerWrite ? "w" : "-"
        str += ownerExecute ? "x" : "-"
        str += groupRead ? "r" : "-"
        str += groupWrite ? "w" : "-"
        str += groupExecute ? "x" : "-"
        str += otherRead ? "r" : "-"
        str += otherWrite ? "w" : "-"
        str += otherExecute ? "x" : "-"
        return str
    }
    
    static func from(octal: Int) -> FilePermissions {
        FilePermissions(
            ownerRead: octal & 0o400 != 0,
            ownerWrite: octal & 0o200 != 0,
            ownerExecute: octal & 0o100 != 0,
            groupRead: octal & 0o040 != 0,
            groupWrite: octal & 0o020 != 0,
            groupExecute: octal & 0o010 != 0,
            otherRead: octal & 0o004 != 0,
            otherWrite: octal & 0o002 != 0,
            otherExecute: octal & 0o001 != 0
        )
    }
}

File: WindowsClient/FileExploder/Models/RemoteFile.cs
using FileExploder.Utilities;

namespace FileExploder.Models;

/// One entry from a directory listing.
///
/// Identity (see <see cref="Id"/>) is deliberately narrower than structural
/// equality: two listings of the same path are "the same row" for
/// selection/drag-and-drop purposes even if, say, its size changed between
/// an auto-refresh tick and the one before it.
public sealed class RemoteFile : IEquatable<RemoteFile>
{
    public string Name { get; }
    public string Path { get; }
    public long Size { get; }
    public DateTimeOffset ModificationDate { get; }
    public bool IsDirectory { get; }
    public bool IsSymlink { get; }
    public FilePermissions Permissions { get; }

    public RemoteFile(
        string name,
        string path,
        long size,
        DateTimeOffset modificationDate,
        bool isDirectory,
        bool isSymlink,
        FilePermissions permissions)
    {
        Name = name;
        Path = path;
        Size = size;
        ModificationDate = modificationDate;
        IsDirectory = isDirectory;
        IsSymlink = isSymlink;
        Permissions = permissions;
    }

    /// Row identity for selection, drag-and-drop and lookups - this is
    /// `Path` itself, not a re-encoded form of it.
    ///
    /// The Swift client hex-encodes this path's UTF-8 bytes instead, because
    /// Swift's default String equality treats canonically-equivalent Unicode
    /// spellings (an NFC and an NFD form of the same visible name) as equal,
    /// while a Linux server keeps them as two genuinely different files - a
    /// directory shared with a Mac routinely ends up holding both, and every
    /// identity check needs to keep them apart. .NET's default string
    /// equality is ordinal (byte-for-byte on the UTF-16 code units, no
    /// Unicode normalization), so an NFC and an NFD spelling already compare
    /// as different strings here - the platform gives this guarantee for
    /// free, without needing the same workaround.
    public string Id => Path;

    public string FormattedSize => IsDirectory ? "--" : FormatUtils.FormattedSize(Size);
    public string FormattedDate => FormatUtils.FormattedDate(ModificationDate);
    public string SymbolicPermissions => Permissions.SymbolicString(IsDirectory, IsSymlink);

    public bool Equals(RemoteFile? other) => other is not null && string.Equals(Id, other.Id, StringComparison.Ordinal);
    public override bool Equals(object? obj) => Equals(obj as RemoteFile);
    public override int GetHashCode() => StringComparer.Ordinal.GetHashCode(Id);
}

public sealed record FilePermissions(
    bool OwnerRead,
    bool OwnerWrite,
    bool OwnerExecute,
    bool GroupRead,
    bool GroupWrite,
    bool GroupExecute,
    bool OtherRead,
    bool OtherWrite,
    bool OtherExecute)
{
    public string SymbolicString(bool isDirectory, bool isSymlink)
    {
        var kind = isSymlink ? 'l' : isDirectory ? 'd' : '-';
        Span<char> chars = stackalloc char[10];
        chars[0] = kind;
        chars[1] = OwnerRead ? 'r' : '-';
        chars[2] = OwnerWrite ? 'w' : '-';
        chars[3] = OwnerExecute ? 'x' : '-';
        chars[4] = GroupRead ? 'r' : '-';
        chars[5] = GroupWrite ? 'w' : '-';
        chars[6] = GroupExecute ? 'x' : '-';
        chars[7] = OtherRead ? 'r' : '-';
        chars[8] = OtherWrite ? 'w' : '-';
        chars[9] = OtherExecute ? 'x' : '-';
        return new string(chars);
    }

    public static FilePermissions FromOctal(int octal) => new(
        OwnerRead: (octal & 0b100_000_000) != 0,
        OwnerWrite: (octal & 0b010_000_000) != 0,
        OwnerExecute: (octal & 0b001_000_000) != 0,
        GroupRead: (octal & 0b000_100_000) != 0,
        GroupWrite: (octal & 0b000_010_000) != 0,
        GroupExecute: (octal & 0b000_001_000) != 0,
        OtherRead: (octal & 0b000_000_100) != 0,
        OtherWrite: (octal & 0b000_000_010) != 0,
        OtherExecute: (octal & 0b000_000_001) != 0);
}
